src/toil/batchSystems/awsBatch.py

```diff
@@ -0,0 +1,519 @@
+
+                    # Compose a result
+                    result = UpdatedBatchJobInfo(jobID=bs_id, exitStatus=exit_code, wallTime=runtime, exitReason=exit_reason)
+
+                    # No more iteration needed, we found a result.
+                    break
+
+            # After the iteration, drop all the records for tasks we acknowledged
+            for (aws_id, bs_id) in acknowledged:
+                del self.aws_id_to_bs_id[aws_id]
+                del self.bs_id_to_aws_id[bs_id]
+                if aws_id in self.killed_job_aws_ids:
+                    # We don't need to remember that we killed this job anymore.
+                    self.killed_job_aws_ids.remove(aws_id)
+
+            if not maxWait:
+                # Don't wait at all
+                break
+            elif result is None:
+                # Wait a bit and poll again
+                time.sleep(min(maxWait/2, 1.0))
+
+        # When we get here we have all the result we can get
+        return result
+
+    def shutdown(self) -> None:
+
+        # Shutdown local processes first
+        self.shutdownLocal()
+
+        for aws_id in self.aws_id_to_bs_id.keys():
+            # Shut down all the AWS jobs we issued.
+            self.__try_terminate(aws_id)
+
+        # Get rid of the job definition we are using if we can.
+        self.__destroy_job_definition()
+
+    @retry(errors=[BotoServerError])
+    def __try_terminate(self, aws_id: str) -> None:
+        """
+        Try to terminate an AWS Batch job.
+
+        Succeed if it can't be canceled because it has stopped,
+        but fail if it can't be canceled for some other reason.
+        """
+        # Remember that we killed this job so we don't show it as updated
+        # later.
+        self.killed_job_aws_ids.add(aws_id)
+        # Kill the AWS Batch job
+        self.client.terminate_job(jobId=aws_id, reason='Killed by Toil')
+
+    @retry(errors=[BotoServerError])
+    def __wait_until_stopped(self, aws_id: str) -> None:
+        """
+        Wait for a terminated job to actually stop. The AWS Batch API does not
+        guarantee that the status of a job will be SUCCEEDED or FAILED as soon
+        as a terminate call succeeds for it, but Toil requires that a job that
+        has been successfully killed can no longer be observed to be running.
+        """
+
+        while True:
+            # Poll the job
+            response = self.client.describe_jobs(jobs=[aws_id])
+            jobs = response.get('jobs', [])
+            if len(jobs) == 0:
+                # Job no longer exists at all
+                return
+            job = jobs[0]
+            if job.get('status') and job['status'] in ['SUCCEEDED', 'FAILED']:
+                # The job has stopped
+                return
+            # Otherwise the job is still going. Wait for it to stop.
+            logger.info('Waiting for killed job %s to stop', self.aws_id_to_bs_id.get(aws_id, aws_id))
+            time.sleep(2)
```

Comment:
Should this have a timeout?  Also... in the case of long running processes (i.e. week+), we may want to include additional retries beyond the default, since one may get three errors one day and three more the next, etc. if it's intended to run that long.

Reply:
This is only used on killed jobs, so long-running jobs aren't a concern here.

I also don't think it can have a timeout, unless we add a system to remember jobs that are supposed to be dead, and filter them from our view of the jobs that exist/are running. `killBatchJobs` isn't allowed to return until the jobs it was sent to kill appear dead, and if we don't wait until they are well and truly dead on the AWS side we'll have to fake it.

If we want to protect the user from a stuck job on the AWS side, we could determine a time that it's reasonable to wait on AWS to kill the job, and crash Toil if it is exceeded.